bitcoin/testBwm.c

```diff
@@ -0,0 +1,1063 @@
+    record->manager= manager;
+    record->wallet= wallet;
+    record->transaction = transaction;
+    record->event = event;
+    pthread_mutex_lock (&state->lock);
+    array_add (state->transactionEvents, record);
+    pthread_mutex_unlock (&state->lock);
+}
+
+static void
+_testWalletEventRecordingCallback (BRWalletManagerClientContext context,
+                                   BRWalletManager manager,
+                                   BRWallet *wallet,
+                                   BRWalletEvent event) {
+    BRRunTestWalletManagerSyncState *state = (BRRunTestWalletManagerSyncState*) context;
+    struct BRWalletEventRecord *record = calloc (1, sizeof (struct BRWalletEventRecord));
+    record->manager= manager;
+    record->wallet= wallet;
+    record->event = event;
+
+    pthread_mutex_lock (&state->lock);
+    array_add (state->walletEvents, record);
+    pthread_mutex_unlock (&state->lock);
+}
+
+static void
+_testWalletManagerEventRecordingCallback (BRWalletManagerClientContext context,
+                                          BRWalletManager manager,
+                                          BRWalletManagerEvent event) {
+    BRRunTestWalletManagerSyncState *state = (BRRunTestWalletManagerSyncState*) context;
+    struct BRWalletManagerEventRecord *record = calloc (1, sizeof (struct BRWalletManagerEventRecord));
+    record->manager= manager;
+    record->event = event;
+
+    pthread_mutex_lock (&state->lock);
+    array_add (state->managerEvents, record);
+    pthread_mutex_unlock (&state->lock);
+}
+
+typedef struct {
+    uint8_t kill;
+    BRWalletManager manager;
+} BRRunTestWalletManagerSyncThreadState;
+
+static void *
+_testBRWalletManagerConnectThread (void *context) {
+    BRRunTestWalletManagerSyncThreadState *state = (BRRunTestWalletManagerSyncThreadState *) context;
+    while (!state->kill) {
+        BRWalletManagerConnect (state->manager);
+    }
+    return NULL;
+}
+
+static void *
+_testBRWalletManagerDisconnectThread (void *context) {
+    BRRunTestWalletManagerSyncThreadState *state = (BRRunTestWalletManagerSyncThreadState *) context;
+    while (!state->kill) {
+        BRWalletManagerDisconnect (state->manager);
+    }
+    return NULL;
+}
+
+static void *
+_testBRWalletManagerScanThread (void *context) {
+    BRRunTestWalletManagerSyncThreadState *state = (BRRunTestWalletManagerSyncThreadState *) context;
+    while (!state->kill) {
+        BRWalletManagerScan (state->manager);
+    }
+    return NULL;
+}
+
+static void
+BRRunTestWalletManagerSyncTestSetup (BRRunTestWalletManagerSyncState *state, uint64_t blockHeight) {
+    state->blockHeight = blockHeight;
+    array_new (state->managerEvents, 1);
+    array_new (state->walletEvents, 1);
+    array_new (state->transactionEvents, 1);
+    pthread_mutex_init (&state->lock, NULL);
+}
+
+static int
+BRRunTestWalletManagerSyncTestVerification (BRRunTestWalletManagerSyncState *state) {
```

Comment:
This is checking that that events that did occur occurred in an allowed order but not that any particular event occurred?  Thus, you could have success with an event sequence of [WALLET_MANAGER_CREATED] only - w/o ever seeing WALLET_MANAGER_WALLET_ADDED?

Reply:
That is correct. I can add to the routine to confirm that WALLET_MANAGER_CREATED was received, and WALLET_MANAGER_WALLET_ADDED thereafter.